### PR TITLE
Replace coveralls with codecov.io

### PR DESCRIPTION
**Pull request by crusaderky**
As discussed in a recent maintainer's meeting, coveralls interacts poorly with github actions, whereas codecov.io offers a much smoother experience.

@jrbourbeau I need support from an admin to integrate codecov.io with the repo
@jsignell FYI this interacts with your PR #7198

### CODE REVIEW

File: continuous_integration/scripts/run_tests.sh [hunk 1]
```diff
@@ -10,8 +10,8 @@ if [[ $PARALLEL == 'true' ]]; then
 fi
 
 if [[ $COVERAGE == 'true' ]]; then
-    echo "coverage run `which py.test` dask --runslow --doctest-modules $XTRATESTARGS"
-    coverage run `which py.test` dask --runslow --doctest-modules $XTRATESTARGS
+    echo "coverage run -m pytest dask --runslow --doctest-modules $XTRATESTARGS"
+    coverage run -m pytest dask --runslow --doctest-modules $XTRATESTARGS
```
Comment (jrbourbeau) [inline continuous_integration/scripts/run_tests.sh:14]: We moved doctests to a separate GH actions build (xref https://github.com/dask/dask/pull/7238)

```suggestion
    echo "coverage run -m pytest dask --runslow $XTRATESTARGS"
    coverage run -m pytest dask --runslow $XTRATESTARGS
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -14,7 +14,7 @@ jobs:
     env:
       PYTHON_VERSION: ${{ matrix.python-version }}
       PARALLEL: "true"
-      COVERAGE: "false"
+      COVERAGE: "true"
```
Comment (jrbourbeau) [inline .github/workflows/ci.yml:17]: Does coverage work when running tests in parallel?

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the ping @crusaderky, I'll look into setting up codecov

Also cc @jacobtomlinson who may find this interesting

**Comment by jrbourbeau** [comment]
From looking through https://github.com/codecov/codecov-action#usage, it looks like codecov supports tokenless uploads for public repos. I'm curious if adding the codecov action step will "just work"

**Comment by crusaderky** [comment]
> From looking through https://github.com/codecov/codecov-action#usage, it looks like codecov supports tokenless uploads for public repos. I'm curious if adding the codecov action step will "just work"

I tested it extensively in all of my repos; no token needed

File: .github/workflows/ci.yml (hunk 2 above)
Comment (crusaderky) [reply .github/workflows/ci.yml:17]: Whoops it doesn't out of the box. Should be fixed now.

**Comment by crusaderky** [comment]
@jrbourbeau ready for final review and merge

**Review by jrbourbeau — COMMENTED** [review]
Overall this looks great, thanks for your work on this @crusaderky. I noticed there [are some coverage-related warnings in CI](https://github.com/dask/dask/pull/7246/checks?check_run_id=1951919278#step:6:1071) along the lines of

```
Coverage.py warning: Data file '/home/runner/work/dask/dask/.coverage.fv-az148-411.11930.085848' doesn't seem to be a coverage data file: cannot unpack non-iterable NoneType object
Coverage.py warning: Data file '/home/runner/work/dask/dask/.coverage.fv-az148-411.10659.561995' doesn't seem to be a coverage data file: Couldn't use data file '/home/runner/work/dask/dask/.coverage.fv-az148-411.10659.561995': no such table: coverage_schema
```

Do you have any thoughts on what might be causing our coverage reports to trigger these warnings?

**Comment by crusaderky** [comment]
> Do you have any thoughts on what might be causing our coverage reports to trigger these warnings?

I suspect they may be created by the workers of a LocalCluster.
The main pytest process spawns 4 subprocesses as test workers (and that's well supported by pytest-coverage); then each of those 4 subprocesses spawns its own subprocesses whenever it fires up the workers of a LocalCluster. That I suspect is catching pytest-coverage unprepared.

Ultimately they are harmless noise, unless we fancy verifying the coverage of a function that runs remotely on the distributed scheduler - which is important for the distributed package, but not so much here.

**Comment by jrbourbeau** [comment]
Alright, let's give this a shot then

**Comment by jrbourbeau** [comment]
https://app.codecov.io/gh/dask/dask/ looks good 🎉

**Comment by jacobtomlinson** [comment]
Awesome stuff! I'm a big fan of codecov.

I wonder if there is any reason to have the `COVERAGE` env var switch any more, codecov handles merging of coverage reports from multiple runs.

**Pull request merged**